### PR TITLE
Improve accuracy for Geo Centroid Aggregation

### LINKED ISSUE #41032: Improve accuracy for Geo Centroid Aggregation
`Geo Centroid aggregation` accuracy decreases drastically with the number of points being aggregated. This is due to the current implementation that uses a moving average and encodes and decodes every partial result.

Accuracy can be improved by not encoding/decoding such results.

Derived from the issue reported here: https://discuss.elastic.co/t/inaccuracy-for-geo-centroid-aggregation-and-geo-bounds-aggregation/175839/1

### PR DESCRIPTION
**Pull request by iverase**
keep the partial results as doubles and uses Kahan summation to help reduce floating point errors.

closes #41032

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalGeoCentroid.java [hunk 1]
```diff
@@ -81,8 +66,8 @@ protected void doWriteTo(StreamOutput out) throws IOException {
         out.writeVLong(count);
         if (centroid != null) {
             out.writeBoolean(true);
-            // should we just write lat and lon separately?
-            out.writeLong(encodeLatLon(centroid.lat(), centroid.lon()));
+            out.writeDouble(centroid.lat());
+            out.writeDouble(centroid.lon());
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalGeoCentroid.java]: and there

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalGeoCentroid.java [hunk 2]
```diff
@@ -69,8 +55,7 @@ public InternalGeoCentroid(StreamInput in) throws IOException {
         super(in);
         count = in.readVLong();
         if (in.readBoolean()) {
-            final long hash = in.readLong();
-            centroid = new GeoPoint(decodeLatitude(hash), decodeLongitude(hash));
+            centroid = new GeoPoint(in.readDouble(), in.readDouble());
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalGeoCentroid.java]: we need to handle backward compatibility here

**Review by jpountz — CHANGES_REQUESTED** [review]
The serialization logic needs to be backward compatible. Otherwise it looks good.

**Comment by iverase** [comment]
Thanks @jpountz. Of course you might be talking to nodes with different versions. I wonder if it is then worthy to change that logic.

**Comment by jpountz** [comment]
@iverase Actually I was suggesting that we change the serialization format based on the Stream(In|Out)put#getVersion(). If the version is gte 7.1 then we would serialize pairs of doubles and otherwise we would keep the old that folds both lat and lon into a long.

**Comment by iverase** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Comment by iverase** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample please

**Comment by polyfractal** [comment]
There aren't any yaml tests for geo centroid, right?  Just double checking since those can run in mixed-cluster BWC tests and might fail before the backport is complete.  I skimmed quickly and didn't see any but wanted to double check :)

**Review by polyfractal — APPROVED** [review]
LGTM!

**Pull request merged**